Don't eagerly evaluate calc() of <integer> values

weinig opened the pull request:
#### 76d664198b75324545784c30904dc1cd2e7d244d
<pre>
Don&apos;t eagerly evaluate calc() of &lt;integer&gt; values
<a href="https://bugs.webkit.org/show_bug.cgi?id=278950">https://bugs.webkit.org/show_bug.cgi?id=278950</a>

Reviewed by Darin Adler.

Makes processing of &lt;integer&gt; primitive values match other
primitive values by deferring calc() evaluation until style
building. With it no longer being eagerly evaluated, we also
now need to handle it as a potential calc() value to allow
proper clamping to integral bounds.

* LayoutTests/fast/css/order-calculated-value-expected.txt:
* LayoutTests/fast/css/order-calculated-value.html:
* LayoutTests/fast/css/webkit-line-clamp-calculated-value-expected.txt:
* LayoutTests/fast/css/webkit-line-clamp-calculated-value.html:
* LayoutTests/fast/css/z-index-calculated-value-expected.txt:
* LayoutTests/fast/css/z-index-calculated-value.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/signs-abs-computed-expected.txt:
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
(WebCore::CSSCalc::determineCanonicalDimension):
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
(WebCore::CSSCalc::percentageResolveToDimension):
(WebCore::CSSCalc::simplify):
* Source/WebCore/css/calc/CSSCalcType.cpp:
(WebCore::CSSCalc::Type::determinePercentHint):
(WebCore::CSSCalc::Type::matches const):
* Source/WebCore/css/calc/CSSCalcValue.cpp:
(WebCore::CSSCalcValue::primitiveType const):
(WebCore::CSSCalcValue::clampToPermittedRange const):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CSSPrimitiveValueResolver.h:
(WebCore::CSSPropertyParserHelpers::CSSPrimitiveValueResolverBase::resolve):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+IntegerDefinitions.h:
(WebCore::CSSPropertyParserHelpers::IntegerKnownTokenTypeFunctionConsumer::consume):
* Source/WebCore/platform/calc/CalculationCategory.cpp:
(WebCore::Calculation::operator&lt;&lt;):
* Source/WebCore/platform/calc/CalculationCategory.h:

Canonical link: <a href="https://commits.webkit.org/283008@main">https://commits.webkit.org/283008@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6809971945bc8d06e3f87077181d64b0773a78e2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64900 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44265 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17504 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68924 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15506 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67017 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52046 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15785 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52144 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10696 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67966 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40927 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56157 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32764 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37598 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/13552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14382 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59507 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13864 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70629 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8845 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13344 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59470 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8879 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56225 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14328 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7297 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/969 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40077 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/41156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/42337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40897 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/calc/CSSCalcValue.cpp:
```diff
@@ -186,6 +188,9 @@ inline double CSSCalcValue::clampToPermittedRange(double value) const
         return 0;
 
+    if (m_tree.category == Calculation::Category::Integer)
+        value = std::floor(value + 0.5);
```
Isn't there a more efficient way to do this operation?

weinig replied:
Not sure. This is just moving from elsewhere. 

Going to take a look at the assembly and see what the compiler outputs for it verse other rounding functions.

weinig committed:
Don't eagerly evaluate calc() of <integer> values
https://bugs.webkit.org/show_bug.cgi?id=278950

Reviewed by Darin Adler.

Makes processing of <integer> primitive values match other
primitive values by deferring calc() evaluation until style
building. With it no longer being eagerly evaluated, we also
now need to handle it as a potential calc() value to allow
proper clamping to integral bounds.

* LayoutTests/fast/css/order-calculated-value-expected.txt:
* LayoutTests/fast/css/order-calculated-value.html:
* LayoutTests/fast/css/webkit-line-clamp-calculated-value-expected.txt:
* LayoutTests/fast/css/webkit-line-clamp-calculated-value.html:
* LayoutTests/fast/css/z-index-calculated-value-expected.txt:
* LayoutTests/fast/css/z-index-calculated-value.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/signs-abs-computed-expected.txt:
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
(WebCore::CSSCalc::determineCanonicalDimension):
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
(WebCore::CSSCalc::percentageResolveToDimension):
(WebCore::CSSCalc::simplify):
* Source/WebCore/css/calc/CSSCalcType.cpp:
(WebCore::CSSCalc::Type::determinePercentHint):
(WebCore::CSSCalc::Type::matches const):
* Source/WebCore/css/calc/CSSCalcValue.cpp:
(WebCore::CSSCalcValue::primitiveType const):
(WebCore::CSSCalcValue::clampToPermittedRange const):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CSSPrimitiveValueResolver.h:
(WebCore::CSSPropertyParserHelpers::CSSPrimitiveValueResolverBase::resolve):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+IntegerDefinitions.h:
(WebCore::CSSPropertyParserHelpers::IntegerKnownTokenTypeFunctionConsumer::consume):
* Source/WebCore/platform/calc/CalculationCategory.cpp:
(WebCore::Calculation::operator<<):
* Source/WebCore/platform/calc/CalculationCategory.h:

Canonical link: https://commits.webkit.org/283008@main
